File: src/Prompts/Functions/RefactoringSuggestion.php
<?php

declare(strict_types=1);

namespace App\Prompts\Functions;

use App\PromptLoader;
use App\Prompts\Core\CorePrompts;
use App\Prompts\Core\OutputFormats;
use App\Prompts\Enums\Perspective;

/**
 * リファクタリング提案プロンプト
 */
final class RefactoringSuggestion
{
    private static function renderContext(string $context): string
    {
        if ($context === '') {
            return '';
        }

        return PromptLoader::getInstance()->renderTemplate(
            'functions/refactoring-suggestion/context',
            ['context' => $context]
        );
    }

    /**
     * リファクタリング提案プロンプトを生成
     *
     * @param string $code 分析対象コード
     * @param string $context 追加コンテキスト
     * @param string|null $perspective 設計観点（ddd, laravel, clean）
     */
    public static function generate(
        string $code = '',
        string $context = '',
        ?string $perspective = null
    ): string {
        $loader = PromptLoader::getInstance();

        $corePrompt = CorePrompts::all();
        $outputFormat = OutputFormats::forRefactoring();

        $contextSection = self::renderContext($context);

        // 設計観点（Perspective）
        $perspectivePrompt = '';
        if ($perspective !== null) {
            $perspectiveEnum = Perspective::fromAlias($perspective);
            if ($perspectiveEnum !== null) {
                $perspectivePrompt = $loader->getContent($perspectiveEnum->promptPath());
            }
        }

        // 対象コードセクション
        $targetCode = self::buildTargetCodeSection($code);

        return $loader->renderTemplate('functions/refactoring-suggestion/base', [
            'corePrompt' => $corePrompt,
            'perspectivePrompt' => $perspectivePrompt,
            'contextSection' => $contextSection,
            'outputFormat' => $outputFormat,
            'targetCode' => $targetCode,
        ]);
    }

    private static function buildTargetCodeSection(string $code): string
    {
        $loader = PromptLoader::getInstance();

        if ($code === '') {
            return $loader->getContent('functions/refactoring-suggestion/target-code-fallback');
        }

        return $loader->renderTemplate('functions/refactoring-suggestion/target-code', [
            'code' => $code,
        ]);
    }
}

File: internal/functions/refactoring_suggestion.go
package functions

import (
	"seiren/internal/domain"
	"seiren/internal/promptloader"
)

func GenerateRefactoringSuggestion(code, context, perspective string, focuses []string) string {
	loader := promptloader.Get()

	if len(focuses) == 0 {
		focuses = domain.DefaultFocuses()
	}

	resolved := domain.ResolveFocuses(focuses)
	corePrompt := buildPromptFromFocuses(resolved.Core, loader)
	outputFormat := domain.OutputFormatsForRefactoring(loader)

	contextSection := ""
	if context != "" {
		contextSection = loader.RenderTemplate("functions/refactoring-suggestion/context", map[string]string{
			"context": context,
		})
	}

	perspectivePrompt := ""
	if perspective != "" {
		if p, ok := domain.PerspectiveFromAlias(perspective); ok {
			perspectivePrompt = loader.GetContent(p.PromptPath())
		}
	}

	targetCode := buildRefactoringTargetCode(code, loader)

	return loader.RenderTemplate("functions/refactoring-suggestion/base", map[string]string{
		"corePrompt":        corePrompt,
		"perspectivePrompt": perspectivePrompt,
		"contextSection":    contextSection,
		"outputFormat":      outputFormat,
		"targetCode":        targetCode,
	})
}

func buildRefactoringTargetCode(code string, loader *promptloader.Loader) string {
	if code == "" {
		return loader.GetContent("functions/refactoring-suggestion/target-code-fallback")
	}
	return loader.RenderTemplate("functions/refactoring-suggestion/target-code", map[string]string{
		"code": code,
	})
}
